feat(cassandra4): more attributes

FrankSpitulski opened the pull request:
fixes #1298

trask commented:
hey @FrankSpitulski!

we had some discussion recently in https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/602 and https://github.com/open-telemetry/opentelemetry-specification/issues/968, and want to spec even instrumentation-specific attributes

can you submit a PR to the specification repo first? this is probably where I would start: https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/database.md#call-level-attributes-for-specific-technologies

anuraaga commented on instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/MoreAttributes.java:
```diff
@@ -21,4 +21,11 @@
   public static final String HTTP_FRAGMENT = "http.fragment.string";
 
   public static final String USER_NAME = "user.principal";
+  public static final String QUERY_PAGE_SIZE = "query.pageSize";
```
I think in parallel with the spec change we could proceed if this is scoped to cassandra. How about moving these into the cassandra package and prefixing with that for now? @trask what do you think?

trask replied:
> I think in parallel with the spec change we could proceed if this is scoped to cassandra

Once a spec PR is opened, I'm good merging something here that implements the initial spec PR (and we can sync it up later with whatever ends up getting approved/merged in the spec).

> How about moving these into the cassandra package and prefixing with that for now?

👍

iNikem commented:
@FrankSpitulski Are you interested in open a PR in spec repo adding those semantic attributes?

FrankSpitulski commented:
Sure, they tagged the issue I made release:after-ga though so I didn't think it was going to go in.

trask commented:
> Sure, they tagged the issue I made release:after-ga though so I didn't think it was going to go in.

hey @FrankSpitulski, i've asked for clarification https://github.com/open-telemetry/opentelemetry-specification/issues/1058#issuecomment-713327915

FrankSpitulski commented:
@trask should be good now

anuraaga commented on instrumentation/cassandra/cassandra-4.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/cassandra/v4_0/CassandraDatabaseClientTracer.java:
```diff
@@ -49,14 +62,85 @@ protected InetSocketAddress peerAddress(CqlSession cqlSession) {
+  }
+
+  private static boolean isIdempotent(Statement<?> statement) {
+    Boolean idempotent = statement.isIdempotent();
+    if (idempotent != null) {
+      return idempotent;
     }
+    return false;
```
```suggestion
    return Boolean.TRUE.equals(Statement.isIdempotent());
```

anuraaga commented on instrumentation/cassandra/cassandra-4.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/cassandra/v4_0/CassandraDatabaseClientTracer.java:
```diff
@@ -49,14 +62,85 @@ protected InetSocketAddress peerAddress(CqlSession cqlSession) {
+    String tableName = null;
+    Matcher matcher = tableNameRegex.matcher(query);
+    if (matcher.find()) {
+      if (matcher.group(2) != null) {
```
Do we have a unit test for these two types of matches?

FrankSpitulski replied:
I have added unit tests

FrankSpitulski replied:
@anuraaga I can't figure out why the unit test is unable to see the public CassandraTableNameExtractor (and before that, the embedded extractor method). Am I missing some test directives in the gradle file?

FrankSpitulski replied:
for ref the task is :instrumentation:cassandra:cassandra-4.0:javaagent:test

anuraaga replied:
Sorry for the trouble - we merged a change which has our tests behave more like an instrumented app, meaning indeed they don't get access to the agent's classes. You'll need to split out a unit tests module for them (long term we should split out library instrumentation instead).

https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/master/instrumentation/jdbc/javaagent-unittests

FrankSpitulski replied:
should be good now @anuraaga

anuraaga reviewed (commented):
Thanks - small nits

anuraaga reviewed (approved):
Thanks!

FrankSpitulski commented:
@trask @iNikem pinging for review please

trask commented on instrumentation/cassandra/cassandra-4.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/cassandra/v4_0/CassandraTableNameExtractor.java:
```diff
@@ -0,0 +1,38 @@
+
+  private CassandraTableNameExtractor() {}
+
+  private static final Pattern tableNameRegex =
+      Pattern.compile(
+          ".*(?:FROM|INTO|UPDATE|TRUNCATE|(?:CREATE|ALTER|DROP) TABLE)\\s+(?:IF (?:NOT )?EXISTS\\s+)?([A-Z1-9_]+\\.([A-Z1-9_]+)|([A-Z1-9_]+))",
+          Pattern.CASE_INSENSITIVE);
```
from https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/database.md

> It is not recommended to attempt any client-side parsing of `db.statement` just to get this property

We also parse jdbc SQL for the table name, though, violating this, so ??? <-- @mateuszrzeszutek @iNikem @johnbley @anuraaga

If we are going to parse/capture the table name, it would be great to use it in the span name: `<db.operation> <db.name>.<db.sql.table>`

FrankSpitulski replied:
we could extract the table name without parsing anything if the statement is a prepared statement, but if the statement is a simplestatement that's the only way I could find to get it out.

mateuszrzeszutek replied:
We already parse the cassandra query to sanitize it - once #2113 is in operation & table name will be available for free (although the current sanitizer implementation cannot extract table name from create/drop table statements, but I suspect that those are rare cases anyway).

FrankSpitulski replied:
Okay, I'll have a look at refactoring and basing the table extraction on this.

FrankSpitulski replied:
done

trask reviewed (commented):
Thanks for getting this into the specs!!

FrankSpitulski commented on instrumentation/cassandra/cassandra-4.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/cassandra/v4_0/CassandraDatabaseClientTracer.java:
```diff
@@ -49,14 +56,67 @@ protected InetSocketAddress peerAddress(CqlSession cqlSession) {
+  protected Span onConnection(Span span, CqlSession cqlSession) {
+    span = super.onConnection(span, cqlSession);
+    DriverExecutionProfile config = cqlSession.getContext().getConfig().getDefaultProfile();
+    // may be overwritten by statement, but take the default for now
+    int pageSize = config.getInt(DefaultDriverOption.REQUEST_PAGE_SIZE);
+    if (pageSize > 0) {
+      span.setAttribute(SemanticAttributes.DB_CASSANDRA_PAGE_SIZE, pageSize);
+    }
+    // may be overwritten by statement, but take the default for now
+    span.setAttribute(
+        SemanticAttributes.DB_CASSANDRA_CONSISTENCY_LEVEL,
+        config.getString(DefaultDriverOption.REQUEST_CONSISTENCY));
```
that is correct

FrankSpitulski replied:
it's a little less dire than that, on exception it will still attempt to extract the execution info and invoke onResponse

trask replied:
can we avoid capturing the wrong value? In general it's better to capture no value than to capture the wrong value

FrankSpitulski replied:
it seems that actually did not come over on the rebase, I'll add that back in

trask replied:
> it's a little less dire than that, on exception it will still attempt to extract the execution info and invoke onResponse

can you point me where you're seeing that? this is what I'm looking at in `TracingCqlSession`, and on exception it doesn't call `onResponse`:

```
      try {
        ResultSet resultSet = session.execute(query);
        tracer().onResponse(context, resultSet.getExecutionInfo());
        return resultSet;
      } catch (RuntimeException e) {
        tracer().endExceptionally(context, e);
        throw e;
      } finally {
        tracer().end(context);
      }
```

trask replied:
> it seems that actually did not come over on the rebase, I'll add that back in

👍 (ignore my last comment)

FrankSpitulski replied:
what is the best way to clear the attribute on a span @trask? The setAttribute method is annotated with not null.

trask replied:
can we set the attributes only once we have the statement?

FrankSpitulski replied:
the default values comes out of the session which is only available in onConnection, not in onResponse

trask replied:
can we pass the session in to `onResponse`?

FrankSpitulski replied:
we can, but it means the endExceptionally method no longer overrides the superclass method since it also needs the session. I have added that an implementation that throws an illegal state exception if called.

another option would be to pass the session and execution info as a pair, but that may create undesired garbage.

FrankSpitulski replied:
hey @trask, are you happy with the change to add the session into onResponse meaning it also needs to be passed into endExceptionally?

FrankSpitulski commented on instrumentation/cassandra/cassandra-4.0/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/cassandra/v4_0/CassandraDatabaseClientTracer.java:
```diff
@@ -49,14 +56,67 @@ protected InetSocketAddress peerAddress(CqlSession cqlSession) {
+    if (statement.getPageSize() > 0) {
+      span.setAttribute(SemanticAttributes.DB_CASSANDRA_PAGE_SIZE, statement.getPageSize());
+    }
+    span.setAttribute(SemanticAttributes.DB_CASSANDRA_IDEMPOTENCE, isIdempotent(statement));
```
statements are not idempotent by default. I can't think of a situation where it would be idempotent but not tagged as such. by default it isn't specified though so null should be treated as false. for example, I just tested out running an insert statement which is not idempotent. it had a null idempotent field.

trask replied:
the javadoc for isIdempotent() says it returns `null to use the default value defined in the configuration` and references DefaultDriverOption.REQUEST_DEFAULT_IDEMPOTENCE

FrankSpitulski replied:
hmm, that's a good catch. I will add it into the same set as the other default driver options.

trask reviewed (approved):
@FrankSpitulski thanks for your patience with us, and great work again driving this through the specification process!

iNikem commented:
@FrankSpitulski can you please rebase this PR? Then it can finally be merged. Sorry for such slow progress :(

FrankSpitulski commented:
@iNikem It's more than a rebase after the changes in #2113. I'll rewrite it to fit as soon as I can.

FrankSpitulski commented:
@iNikem I have rebased and updated the code to use the new SqlStatementSanitizer. As @mateuszrzeszutek mentioned it can no longer handle CREATE statements. It also no longer splits out the table name from the keyspace when the statement has the form <keyspace>.<table>, which I believe should be considered a bug in the SqlStatementSanitizer.

anuraaga commented:
Let's go ahead and get this in - @mateuszrzeszutek do you mind following up with the SQL issues?

anuraaga commented:
@FrankSpitulski Thanks a lot for this contribution and bearing with the long review process! 🙏

mateuszrzeszutek commented:
> it can no longer handle CREATE statements

I'll create an issue for that 👍 

>  It also no longer splits out the table name from the keyspace when the statement has the form `<keyspace>.<table>`, which I believe should be considered a bug in the SqlStatementSanitizer.

Actually, this is intentional: spec says that `db.sql.table` is

> the name of the primary table that the operation is acting upon, including the schema name (if applicable).

so whatever sanitizer extracts as "table" may contain the schema name.
I assume that in Cassandra's case keyspace is equivalent to schema name. I think that the best course of action would be to just split the table string in Cassandra instrumentation.

trask commented:
🎉 thank you @FrankSpitulski
